File: flowspot-omni-src-0.8/src/main/scala/omni/scan/OptimizationConfig.scala
package omni.scan

/**
 * 扫描优化配置类
 * 控制数据流分析中的各种去重和过滤选项
 */
case class OptimizationConfig(
  // 子路径去重：移除被其他路径包含的子路径，保留更完整的数据流路径
  enableSubPathDeduplication: Boolean = true,
  
  // Sink位置去重：基于sink的文件位置进行精确去重，避免相同位置的重复报告
  enableSinkLocationDeduplication: Boolean = true,
  
  // 上下文过滤：应用现有的上下文过滤器，移除特定上下文中的误报
  enableContextFiltering: Boolean = true
) {
  
  /**
   * 获取配置的简短描述
   */
  def getDescription: String = {
    val enabled = List(
      if (enableSubPathDeduplication) Some("子路径去重") else None,
      if (enableSinkLocationDeduplication) Some("位置去重") else None,
      if (enableContextFiltering) Some("上下文过滤") else None
    ).flatten
    
    if (enabled.nonEmpty) {
      s"已启用: ${enabled.mkString(", ")}"
    } else {
      "所有优化已禁用"
    }
  }
  
  /**
   * 获取详细的配置说明
   */
  def getDetailedDescription: String = {
    s"""扫描优化配置:
       |• 子路径去重: ${if (enableSubPathDeduplication) "启用" else "禁用"} - 移除被包含的短路径，保留完整数据流
       |• Sink位置去重: ${if (enableSinkLocationDeduplication) "启用" else "禁用"} - 基于文件位置去重，避免重复报告
       |• 上下文过滤: ${if (enableContextFiltering) "启用" else "禁用"} - 应用上下文过滤器，减少误报
       |""".stripMargin
  }
}

object OptimizationConfig {
  /**
   * 默认配置：所有优化都启用
   */
  val default: OptimizationConfig = OptimizationConfig()
  
  /**
   * 无优化配置：所有优化都禁用，用于调试或特殊需求
   */
  val noOptimization: OptimizationConfig = OptimizationConfig(
    enableSubPathDeduplication = false,
    enableSinkLocationDeduplication = false,
    enableContextFiltering = false
  )
  
  /**
   * 轻量优化配置：只启用基本的位置去重
   */
  val lightOptimization: OptimizationConfig = OptimizationConfig(
    enableSubPathDeduplication = false,
    enableSinkLocationDeduplication = true,
    enableContextFiltering = false
  )
}

File: flowspot-plugin/src/main/java/com/flowspot/intellij/model/OptimizationConfig.java
/*
 * Copyright 2024 FlowSpot plugin contributors
 *
 * This file is part of IntelliJ FlowSpot plugin.
 *
 * IntelliJ FlowSpot plugin is free software: you can redistribute it 
 * and/or modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation, either version 3 of 
 * the License, or (at your option) any later version.
 *
 * IntelliJ FlowSpot plugin is distributed in the hope that it will
 * be useful, but WITHOUT ANY WARRANTY; without even the implied 
 * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with IntelliJ FlowSpot plugin.
 * If not, see <http://www.gnu.org/licenses/>.
 */
package com.flowspot.intellij.model;

/**
 * FlowSpot 优化配置类
 * 用于配置数据流分析的优化选项
 */
public class OptimizationConfig {
    
    private final boolean enableSubPathDeduplication;
    private final boolean enableSinkLocationDeduplication;
    private final boolean enableContextFiltering;
    
    /**
     * 构造函数
     * 
     * @param enableSubPathDeduplication 是否启用子路径去重
     * @param enableSinkLocationDeduplication 是否启用Sink位置去重
     * @param enableContextFiltering 是否启用上下文过滤
     */
    public OptimizationConfig(boolean enableSubPathDeduplication,
                            boolean enableSinkLocationDeduplication,
                            boolean enableContextFiltering) {
        this.enableSubPathDeduplication = enableSubPathDeduplication;
        this.enableSinkLocationDeduplication = enableSinkLocationDeduplication;
        this.enableContextFiltering = enableContextFiltering;
    }
    
    /**
     * 创建默认配置（所有优化都启用）
     */
    public static OptimizationConfig createDefault() {
        return new OptimizationConfig(true, true, true);
    }
    
    /**
     * 是否启用子路径去重
     */
    public boolean isEnableSubPathDeduplication() {
        return enableSubPathDeduplication;
    }
    
    /**
     * 是否启用Sink位置去重
     */
    public boolean isEnableSinkLocationDeduplication() {
        return enableSinkLocationDeduplication;
    }
    
    /**
     * 是否启用上下文过滤
     */
    public boolean isEnableContextFiltering() {
        return enableContextFiltering;
    }
    
    /**
     * 转换为Scala项目中的OptimizationConfig
     * 这个方法将在传递给Scala项目时使用
     */
    public omni.scan.OptimizationConfig toScalaConfig() {
        return new omni.scan.OptimizationConfig(
            enableSubPathDeduplication,
            enableSinkLocationDeduplication,
            enableContextFiltering
        );
    }
    
    @Override
    public String toString() {
        return "OptimizationConfig{" +
                "enableSubPathDeduplication=" + enableSubPathDeduplication +
                ", enableSinkLocationDeduplication=" + enableSinkLocationDeduplication +
                ", enableContextFiltering=" + enableContextFiltering +
                '}';
    }
    
    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        
        OptimizationConfig that = (OptimizationConfig) o;
        
        if (enableSubPathDeduplication != that.enableSubPathDeduplication) return false;
        if (enableSinkLocationDeduplication != that.enableSinkLocationDeduplication) return false;
        return enableContextFiltering == that.enableContextFiltering;
    }
    
    @Override
    public int hashCode() {
        int result = (enableSubPathDeduplication ? 1 : 0);
        result = 31 * result + (enableSinkLocationDeduplication ? 1 : 0);
        result = 31 * result + (enableContextFiltering ? 1 : 0);
        return result;
    }
}
